bug: `ModalOptions` no longer accepts a Type parameter

### Prerequisites

- [x] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/docs/CONTRIBUTING.md#creating-an-issue).
- [x] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [x] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v8.x

### Current Behavior

Using this code:

```ts
import { ModalOptions } from '@ionic/angular';

const modal = await this.modalController.create({
  component: MyModalComponent,
} satisfies ModalOptions<MyModalComponent>);
```

It errors:
```
[ERROR] TS2315: Type 'ModalOptions' is not generic.
```

### Expected Behavior

This code:

```ts
import { ModalOptions } from '@ionic/angular';

const modal = await this.modalController.create({
  component: MyModalComponent,
} satisfies ModalOptions<MyModalComponent>);
```

Should not error and correctly verify that `component` is `MyModalComponent`

### Steps to Reproduce

N/A

### Code Reproduction URL

N/A

### Ionic Info

N/A

### Additional Information

Bug introduced in https://github.com/ionic-team/ionic-framework/pull/30899, specifically in commit 641f3194ad040bb7730243f645dcc5f3cd886249.

Ionic 8.8.0 was the first version with this bug.

--- a/packages/angular/common/src/types/overlay-options.ts
+++ b/packages/angular/common/src/types/overlay-options.ts
@@ -1,11 +1,15 @@
 import type { Injector } from '@angular/core';
-import type { ModalOptions as CoreModalOptions, PopoverOptions as CorePopoverOptions } from '@ionic/core/components';
+import type {
+  ComponentRef,
+  ModalOptions as CoreModalOptions,
+  PopoverOptions as CorePopoverOptions,
+} from '@ionic/core/components';
 
 /**
  * Modal options with Angular-specific injector support.
  * Extends @ionic/core ModalOptions with an optional injector property.
  */
-export type ModalOptions = CoreModalOptions & {
+export type ModalOptions<T extends ComponentRef = ComponentRef> = CoreModalOptions<T> & {
   injector?: Injector;
 };
 
@@ -13,6 +17,6 @@ export type ModalOptions = CoreModalOptions & {
  * Popover options with Angular-specific injector support.
  * Extends @ionic/core PopoverOptions with an optional injector property.
  */
-export type PopoverOptions = CorePopoverOptions & {
+export type PopoverOptions<T extends ComponentRef = ComponentRef> = CorePopoverOptions<T> & {
   injector?: Injector;
 };

--- a/packages/angular/test/base/e2e/src/standalone/modal-options-generic.spec.ts
+++ b/packages/angular/test/base/e2e/src/standalone/modal-options-generic.spec.ts
@@ -1,0 +1,24 @@
+import { test, expect } from '@playwright/test';
+
+test.describe('Modal Options: Generic Type Parameter', () => {
+  test.beforeEach(async ({ page }) => {
+    await page.goto('/standalone/modal-options-generic');
+  });
+
+  test('should open modal created with ModalOptions<typeof Component>', async ({ page }, testInfo) => {
+    testInfo.annotations.push({
+      type: 'issue',
+      description: 'https://github.com/ionic-team/ionic-framework/issues/31012',
+    });
+
+    await page.locator('ion-button#open-modal').click();
+
+    await expect(page.locator('ion-modal')).toBeVisible();
+
+    const greeting = page.locator('#greeting');
+    await expect(greeting).toHaveText('hello world');
+
+    await page.locator('#close-modal').click();
+    await expect(page.locator('ion-modal')).not.toBeVisible();
+  });
+});

--- a/packages/angular/test/base/src/app/standalone/app-standalone/app.routes.ts
+++ b/packages/angular/test/base/src/app/standalone/app-standalone/app.routes.ts
@@ -24,6 +24,7 @@ export const routes: Routes = [
       },
       { path: 'programmatic-modal', loadComponent: () => import('../programmatic-modal/programmatic-modal.component').then(c => c.ProgrammaticModalComponent) },
       { path: 'modal-custom-injector', loadComponent: () => import('../modal-custom-injector/modal-custom-injector.component').then(c => c.ModalCustomInjectorComponent) },
+      { path: 'modal-options-generic', loadComponent: () => import('../modal-options-generic/modal-options-generic.component').then(c => c.ModalOptionsGenericComponent) },
       { path: 'popover-custom-injector', loadComponent: () => import('../popover-custom-injector/popover-custom-injector.component').then(c => c.PopoverCustomInjectorComponent) },
       { path: 'router-outlet', loadComponent: () => import('../router-outlet/router-outlet.component').then(c => c.RouterOutletComponent) },
       { path: 'back-button', loadComponent: () => import('../back-button/back-button.component').then(c => c.BackButtonComponent) },

--- a/packages/angular/test/base/src/app/standalone/home-page/home-page.component.html
+++ b/packages/angular/test/base/src/app/standalone/home-page/home-page.component.html
@@ -125,6 +125,11 @@
         Modal Custom Injector Test
       </ion-label>
     </ion-item>
+    <ion-item routerLink="/standalone/modal-options-generic">
+      <ion-label>
+        Modal Options Generic Test
+      </ion-label>
+    </ion-item>
     <ion-item routerLink="/standalone/overlay-controllers">
       <ion-label>
         Overlay Controllers Test

--- a/packages/angular/test/base/src/app/standalone/modal-options-generic/modal-options-generic.component.ts
+++ b/packages/angular/test/base/src/app/standalone/modal-options-generic/modal-options-generic.component.ts
@@ -1,0 +1,38 @@
+import { Component, inject } from '@angular/core';
+import { IonContent, IonHeader, IonTitle, IonToolbar, IonButton, ModalController } from '@ionic/angular/standalone';
+import type { ModalOptions } from '@ionic/angular/standalone';
+import { GenericModalComponent } from './modal/modal.component';
+
+@Component({
+  selector: 'app-modal-options-generic',
+  template: `
+    <ion-header>
+      <ion-toolbar>
+        <ion-title>Modal Options Generic Test</ion-title>
+      </ion-toolbar>
+    </ion-header>
+    <ion-content class="ion-padding">
+      <ion-button id="open-modal" (click)="openModal()">
+        Open Modal
+      </ion-button>
+    </ion-content>
+  `,
+  standalone: true,
+  imports: [IonContent, IonHeader, IonTitle, IonToolbar, IonButton],
+})
+export class ModalOptionsGenericComponent {
+  private modalController = inject(ModalController);
+
+  async openModal() {
+    // Regression: ModalOptions<T> must accept a generic type parameter (#31012)
+    const opts: ModalOptions<typeof GenericModalComponent> = {
+      component: GenericModalComponent,
+      componentProps: {
+        greeting: 'hello world',
+      },
+    };
+
+    const modal = await this.modalController.create(opts);
+    await modal.present();
+  }
+}

--- a/packages/angular/test/base/src/app/standalone/modal-options-generic/modal/modal.component.ts
+++ b/packages/angular/test/base/src/app/standalone/modal-options-generic/modal/modal.component.ts
@@ -1,0 +1,29 @@
+import { Component, Input } from '@angular/core';
+import { IonContent, IonHeader, IonTitle, IonToolbar, IonButton, IonButtons } from '@ionic/angular/standalone';
+
+@Component({
+  selector: 'app-generic-modal',
+  template: `
+    <ion-header>
+      <ion-toolbar>
+        <ion-title>Generic Modal</ion-title>
+        <ion-buttons slot="end">
+          <ion-button id="close-modal" (click)="dismiss()">Close</ion-button>
+        </ion-buttons>
+      </ion-toolbar>
+    </ion-header>
+    <ion-content>
+      <p id="greeting">{{ greeting }}</p>
+    </ion-content>
+  `,
+  standalone: true,
+  imports: [IonContent, IonHeader, IonTitle, IonToolbar, IonButton, IonButtons],
+})
+export class GenericModalComponent {
+  @Input() greeting = '';
+  modal: HTMLIonModalElement | undefined;
+
+  dismiss() {
+    this.modal?.dismiss();
+  }
+}

